src/main/java/org/spongepowered/api/event/map/InitializeMapEvent.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event.map;
+
+import org.spongepowered.api.data.Transaction;
+import org.spongepowered.api.event.Cancellable;
+import org.spongepowered.api.event.Event;
+import org.spongepowered.api.eventgencore.annotation.PropertySettings;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.map.MapView;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents an event that is triggered when a {@link org.spongepowered.api.item.ItemTypes#MAP}
+ * is initialized and replaced with a {@link org.spongepowered.api.item.ItemTypes#FILLED_MAP}.
+ */
+public interface InitializeMapEvent extends Event, Cancellable {
+
+    /**
+     * Called before {@link ReplaceMapItem}, by setting {@link #setUsingDefaultBehavior}
+     * in this event, plugins can override the creation of a new map on the disk
+     * and provide customized maps at initialization time. Plugins should use {@link ReplaceMapItem}
+     * to set their own {@link MapView} if overriding vanilla map creation logic.
+     *
+     * <p>Cancelling this event will prevent any {@link MapView} from being created
+     * and will stop {@link ReplaceMapItem} from being fired. This allows for maps
+     * to be disabled or have their initialization overridden by cancelling {@link CreateMapView}.</p>
+     */
+    interface CreateMapView extends InitializeMapEvent {
+
+        /**
+         * Gets if a new {@link MapView} will be created on the disk.
+         *
+         * @return If a new map will be created on the disk
+         */
+        boolean isUsingDefaultBehavior();
```

Comment:
The original value would require allocating it. Thus breaking the purpose
of the event. I want to allow overriding the default with a plugin provided
one. But transaction doesn't allow a null original value.

On Sat, Apr 2, 2016, 2:39 PM Aaron Hill <notifications@github.com> wrote:

> In src/main/java/org/spongepowered/api/event/map/InitializeMapEvent.java
> <https://github.com/SpongePowered/SpongeAPI/pull/1162#discussion_r58298512>
> :
>
> > +     * in this event, plugins can override the creation of a new map on the disk
> > +     * and provide customized maps at initialization time. Plugins should use {@link ReplaceMapItem}
> > +     * to set their own {@link MapView} if overriding vanilla map creation logic.
> > +     *
> > +     * <p>Cancelling this event will prevent any {@link MapView} from being created
> > +     * and will stop {@link ReplaceMapItem} from being fired. This allows for maps
> > +     * to be disabled or have their initialization overridden by cancelling {@link CreateMapView}.</p>
> > +     */
> > +    interface CreateMapView extends InitializeMapEvent {
> > +
> > +        /**
> > +         * Gets if a new {@link MapView} will be created on the disk.
> > +         *
> > +         * @return If a new map will be created on the disk
> > +         */
> > +        boolean isUsingDefaultBehavior();
>
> You should provide the original value too, not just the plugin-set one.
>
> —
> You are receiving this because you authored the thread.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/1162/files/d8338269be828e278d15ed5ffbe9bb30b7c899c1#r58298512>
>